cartographer_ros/cartographer_ros/map_builder_bridge.cc

```diff
@@ -175,4 +175,124 @@ SensorBridge* MapBuilderBridge::sensor_bridge(const int trajectory_id) {
+        map_builder_.sparse_pose_graph()->GetTrajectoryNodes();
+
+    int i = 0;
+    for (const auto& node : trajectory_nodes) {
+      visualization_msgs::Marker marker;
+      marker.id = i++;
+      marker.type = visualization_msgs::Marker::ARROW;
+      marker.header.stamp = ::ros::Time::now();
+      marker.header.frame_id = options_.map_frame;
+      marker.color.b = 1.0;
+      marker.color.a = 1.0;
+      marker.scale.x = 0.1;
+      marker.scale.y = 0.05;
+      marker.scale.z = 0.05;
+      marker.pose = ToGeometryMsgPose(node.pose);
+      trajectory_nodes_list.markers.push_back(marker);
+    }
+  }
+  return trajectory_nodes_list;
+}
+
+std::unordered_map<std::string, visualization_msgs::MarkerArray>
+MapBuilderBridge::GetConstraintsList() {
+  std::unordered_map<std::string, visualization_msgs::MarkerArray>
+      constratins_viz;
+
+  const auto trajectory_nodes =
+      map_builder_.sparse_pose_graph()->GetTrajectoryNodes();
+  const auto constraints = map_builder_.sparse_pose_graph()->constraints();
+  std::vector<cartographer::transform::Rigid3d> submap_transforms;
+  for (auto& trajectory : trajectory_nodes) {
+    auto current_trajectory_transforms =
+        map_builder_.sparse_pose_graph()->GetSubmapTransforms(
+            trajectory.constant_data->trajectory);
+    std::move(current_trajectory_transforms.begin(),
+              current_trajectory_transforms.end(),
+              std::back_inserter(submap_transforms));
+  }
+
+  int id_inter = 0;
+  std::string ns_inter = "inter";
+  int id_intra = 0;
+  std::string ns_intra = "intra";
+  ros::Time now = ros::Time::now();
+  for (const auto& constraint : constraints) {
+    visualization_msgs::Marker constraint_marker, residual_error_marker;
+    int* id_cnt;
+    std::string* tag;
+    std_msgs::ColorRGBA color_constraint, color_error;
+
+    if (constraint.tag ==
+        cartographer::mapping::SparsePoseGraph::Constraint::INTER_SUBMAP) {
+      id_cnt = &id_inter;
+      tag = &ns_inter;
+
+      // yellow
+      color_constraint.a = 1.0;
+      color_constraint.r = color_constraint.g = 1.0;
+
+      // cyan
+      color_error.a = 1.0;
+      color_error.b = color_error.g = 1.0;
+    } else {
+      id_cnt = &id_intra;
+      tag = &ns_intra;
+      color_constraint.a = color_error.a = 1.0;
+      color_constraint.g = 1.0;
+      color_error.b = 1.0;
+    }
+
+    constraint_marker = createVisualizationMarker(
+        (*id_cnt)++, visualization_msgs::Marker::LINE_STRIP, "constraint", now,
+        options_.map_frame);
+    constraint_marker.color = color_constraint;
+    residual_error_marker = createVisualizationMarker(
+        (*id_cnt)++, visualization_msgs::Marker::LINE_STRIP, "residual error",
+        now, options_.map_frame);
+    residual_error_marker.color = color_error;
+
+    geometry_msgs::Point submap_point, submap_pose_point, trajectory_node_point;
+
+    submap_point = ToGeometryMsgPoint(
+        submap_transforms[constraint.submap_id.submap_index]);
```

Comment:
In the current version of the code, you can use `SparsePoseGraph::GetSubmapTransforms` with the `trajectory_id` to get the submap poses of the trajectory ID this submap belongs to, then access the correct submap of this trajectory using the `submap_index`.